src/vdaf/mastic.rs

```diff
@@ -379,32 +391,485 @@ where
+            MasticPrepareMessage::LaterRound => Some(0),
+        }
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareMessage<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match prep_state {
+            MasticPrepareState::FirstRound(_, _, verifier_len, requires_joint_rand) => {
+                Ok(MasticPrepareMessage::FirstRound(
+                    SzkVerifier::decode_with_param(&(*requires_joint_rand, *verifier_len), bytes)?,
+                ))
+            }
+            MasticPrepareState::LaterRound(_, _) => Ok(MasticPrepareMessage::LaterRound),
+        }
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Aggregator<SEED_SIZE, NONCE_SIZE> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+
+    /// The type of messages sent by each aggregator at each round of the Prepare Process.
+    ///
+    /// Decoding takes a [`Self::PrepareState`] as a parameter; this [`Self::PrepareState`] may be
+    /// associated with any aggregator involved in the execution of the VDAF.
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+
+    /// Result of preprocessing a round of preparation shares. This is used by all aggregators as an
+    /// input to the next round of the Prepare Process.
+    ///
+    /// Decoding takes a [`Self::PrepareState`] as a parameter; this [`Self::PrepareState`] may be
+    /// associated with any aggregator involved in the execution of the VDAF.
+    type PrepareMessage = MasticPrepareMessage<T::Field, SEED_SIZE>;
+
+    /// Begins the Prepare process with the other Aggregators. The [`Self::PrepareState`] returned
+    /// is passed to [`Self::prepare_next`] to get this aggregator's first-round prepare message.
+    ///
+    /// Implements `Vdaf.prep_init` from [VDAF].
+    ///
+    /// [VDAF]: https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-08#section-5.2
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        _agg_id: usize,
+        agg_param: &MasticAggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &MasticPublicShare<VidpfWeight<T::Field>>,
+        input_share: &MasticInputShare<T::Field, SEED_SIZE>,
+    ) -> Result<
+        (
+            MasticPrepareState<T::Field, SEED_SIZE>,
+            MasticPrepareShare<T::Field, SEED_SIZE>,
+        ),
+        VdafError,
+    > {
+        let mut xof = P::init(
+            verify_key,
+            &self.domain_separation_tag(DST_PATH_CHECK_BATCH),
+        );
+        let mut output_shares = Vec::<T::Field>::with_capacity(
+            self.vidpf.weight_parameter * agg_param.level_and_prefixes.prefixes().len(),
```

Comment:
Hmm, should `weight_parameter` just be `weight_len` or something?

Reply:
Terminology is inherited from the previous instantiation of VIDPF; I'm not sure what the other purposes of it were.